src/transform/src/column_knowledge.rs

```diff
@@ -389,6 +430,21 @@ pub struct DatumKnowledge {
 }
 
 impl DatumKnowledge {
+    /// Constructs knowledge for a column that is known to come from an empty
+    /// collection. Useful when initializing knowledge for LetRec bindings.
+    ///
+    /// Note that DatumKnowledge::default() provides the bottom element w.r.t.
+    /// the induced partial DatumKnowledge order (see the field docs).
+    fn empty() -> Self {
+        // Strictly speaking we should evolve DatumKnowledge to an enum with two
+        // variants: Empty and NonEmpty and tweak the `union` operation such
+        // that `Empty union x = x`.
+        Self {
+            value: None,
+            nullable: false, // This is safe because the datum will never exist.
```

Comment:
I think it will be great if we can first workshop in what sense `DatumKnowledge` defines a lattice or even  semi-lattice.

I have not tried to sketch the myselfs, but I'm going to sketch some thoughts:

The `nullable` field is weakened / generalized from `false` to `true`. Consequently, both 
```rust
self.nullable |= other.nullable
```
and
```rust
self.nullable &= other.nullable
```
make sense here because they are are consistent with the definition of a boolean lattice.

The `value` field is `Option<T>`. Two `Option<T>` values `{x, y}` can be weakened by either returning (1) `None` if they differ, or `x` if they are identical. This is consistent with the `union` definition. However, I'm struggling to see how
```rust
if self.value.is_none() {
    self.value = other.value.clone()
}
```
forms a valid meet for `Option<T>`. At the very least this definition it is not symmetric. Consequently, at the moment I'm not quite sure that the `Option<T>` meet always exists, so I'm inclined to argue that `ColumnKnowedge` with its current definition doesn't even form a lattice.

Reply:
I think that for the `value` field we want the flat lattice for the underlying type from slide 18 of this deck: https://www.google.com/url?sa=t&source=web&rct=j&url=https://cs.au.dk/~amoeller/spa/3-lattices-and-fixpoints.pdf&ved=2ahUKEwjM_ZS5zvb9AhVXcPEDHbkAB7wQFnoECBwQAQ&usg=AOvVaw24rv2zsY4lXaSd_7-B1sFj

Reply:
Here is a concrete proposal:

1. I will introduce a new type `Flat` that has `Top`, `Some(Datum)`, and `Bottom` variants to model the flat lattice from my previous comment and use this type for `value` instead of `Option<A>`.
2. I will initialize the `knowledge` for the `LetRec` bindings using the bottom element of the product lattice
    ```rust
    DatumKnowledge {
        value: Bottom,
        nullable: true,
    }
    ```
    which seems reasonable because all of those will be initialized to the empty collection.
2. I will use the correct update function `pref.union(&next)` instead of `prev.absorb(&next)` in the update loop.
3. I will run the update loop for `id.len()` bindings (I can seems like a better heuristic than running it only once).

Reply:
> I think that for the value field we want the flat lattice for the underlying type from [slide 18 of this deck](https://cs.au.dk/~amoeller/spa/3-lattices-and-fixpoints.pdf).

Hmm, you are right that the current `absorb` is not commutative. Yes, changing to the flat lattice seems great! It makes a lot of sense to me, because it's also halfway solving https://github.com/MaterializeInc/materialize/issues/18091: bottom will mean that no value can exist that satisfies our knowledge of the column. We can arrive at this when we either see contradicting conditions, or when we know that the collection is empty. (Top will mean that we don't know anything.)

> I will initialize the knowledge for the LetRec bindings using the bottom element of the product lattice
```
DatumKnowledge {
    value: Bottom,
    nullable: true,
}
```

I think nullable should be `false` in the bottom element.

> I will run the update loop for id.len() bindings (I can seems like a better heuristic than running it only once).

As I noted [here](https://github.com/MaterializeInc/materialize/pull/18323#discussion_r1148009551), it can take more steps than the number of Let bindings to converge, because each binding has multiple columns. Nullability can "spread" between columns: if one column refers to the other in its definition, then it might flip to nullable _after_ the other column flips to nullable. So it can take as much as the total number of columns for the nullable field alone.

Also note that changing to the flat lattice means that the `value` field can also go through two steps. Can the two fields influence each other, i.e., can it happen that, say, the `nullable` field doesn't change in a step, but `value` changes, and then in the next step `nullable` changes? If yes, then it can take as many steps as 3 times the total number of columns to converge in the worst case.

And, as I noted [here](https://github.com/MaterializeInc/materialize/pull/18323#discussion_r1147962554), it is not enough to go upwards on this lattice in a heuristic manner that might not reach the fixpoint, because the final result has to be an _upper_ bound on the ideal result. So we absolutely have to go until fixpoint.